File: C#SrcFiles/FireScript.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using static UnityEngine.RuleTile.TilingRuleOutput;

public class FireScript : MonoBehaviour
{
    
    // This is for the modification of bullet's scale and roatation;
    private Rigidbody2D rb;
    //private GameObject BulletKiller;
    private int attackdamage = 15;

    void Start()
    {
        float bulletSpeed = 20f;
        rb = GetComponent<Rigidbody2D>();
        
        
        float angle = transform.eulerAngles.z; // Get the Z rotation of the object

        // Calculate the X and Y components of a vector using trigonometric functions
        float directionX = Mathf.Cos(angle * Mathf.Deg2Rad);
        float directionY = Mathf.Sin(angle * Mathf.Deg2Rad);

        Vector2 constantDirection = new Vector2(directionX, directionY);

        rb.velocity = constantDirection.normalized * bulletSpeed;

    }

    private void OnTriggerEnter2D(Collider2D collision)
    {

        if (collision.gameObject.CompareTag("BulletKiller"))
        {
            print("Bullet Hit Wall! Deleting bullet!");
            DeleteBullet();
        }
        if (collision.gameObject.CompareTag("Reaper"))
        {
            print("Hit Target");
            // giving 20 damage if bullet hits target;
            collision.gameObject.GetComponent<Reaper>().TakeDamage(attackdamage);
            print("Deleting bullet!");
            DeleteBullet();
        }
        if (collision.gameObject.CompareTag("FireMage"))
        {
            print("Hit Target");
            // giving 20 damage if bullet hits target;
            collision.gameObject.GetComponent<FireMageScript>().TakeDamage(attackdamage-3);
            print("Deleting bullet!");
            DeleteBullet();
        }
        if (collision.gameObject.CompareTag("ScarletWidow"))
        {
            print("Hit Target");
            // giving 20 damage if bullet hits target;
            collision.gameObject.GetComponent<ScarletWidow>().TakeDamage(attackdamage - 5);
            print("Deleting bullet!!");
            DeleteBullet();
        }
        if (collision.gameObject.CompareTag("Cathulu"))
        {
            print("Hit Target");
            // giving 20 damage if bullet hits target;
            collision.gameObject.GetComponent<Cathulu>().TakeDamage(attackdamage - 3);
            print("Deleting bullet!");
            DeleteBullet();
        }
        if (collision.gameObject.CompareTag("Skeleton"))
        {
            print("Hit Target");
            // giving 20 damage if bullet hits target;
            collision.gameObject.GetComponent<SkeletonScript>().TakeDamage(attackdamage - 6);
            print("Deleting bullet!");
            DeleteBullet();
        }
        if (collision.gameObject.CompareTag("LavaGolem"))
        {
            print("Hit Target");
            // giving 20 damage if bullet hits target;
            collision.gameObject.GetComponent<LavaScript>().TakeDamage(attackdamage - 8);
            print("Deleting bullet!");
            DeleteBullet();
        }
    }


    private void DeleteBullet()
    {
        Destroy(gameObject);
    }
}

File: C++SrcFiles/FireScript.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cmath>
#include <algorithm>

using namespace std;

class FireScript : public MonoBehaviour
{
private:
    Rigidbody2D* rb;
    int attackDamage;

public:
    FireScript()
    {
        attackDamage = 15;
    }

    void Start()
    {
        float bulletSpeed = 20.0f;
        rb = GetComponent<Rigidbody2D>();

        // Calculate the direction vector based on the object's rotation
        float angle = transform->eulerAngles.z;
        float directionX = cosf(angle * Mathf::Deg2Rad);
        float directionY = sinf(angle * Mathf::Deg2Rad);
        Vector2 constantDirection = Vector2(directionX, directionY);

        rb->velocity = constantDirection.normalized * bulletSpeed;
    }

    void OnTriggerEnter2D(Collider2D* collision)
    {
        if (collision->gameObject->CompareTag("BulletKiller"))
        {
            cout << "Bullet Hit Wall! Deleting bullet!" << endl;
            DeleteBullet();
        } else if (collision->gameObject->CompareTag("Reaper"))
        {
            collision->gameObject->GetComponent<Reaper>()->TakeDamage(attackDamage);
            DeleteBullet();
        } else if (collision->gameObject->CompareTag("FireMage"))
        {
            collision->gameObject->GetComponent<FireMageScript>()->TakeDamage(attackDamage - 3);
            DeleteBullet();
        } else if (collision->gameObject->CompareTag("ScarletWidow"))
        {
            collision->gameObject->GetComponent<ScarletWidow>()->TakeDamage(attackDamage - 5);
            DeleteBullet();
        } else if (collision->gameObject->CompareTag("Cathulu"))
        {
            collision->gameObject->GetComponent<Cathulu>()->TakeDamage(attackDamage - 3);
            DeleteBullet();
        } else if (collision->gameObject->CompareTag("Skeleton"))
        {
            collision->gameObject->GetComponent<SkeletonScript>()->TakeDamage(attackDamage - 6);
            DeleteBullet();
        } else if (collision->gameObject->CompareTag("LavaGolem"))
        {
            collision->gameObject->GetComponent<LavaScript>()->TakeDamage(attackDamage - 8);
            DeleteBullet();
        }
    }

    void DeleteBullet()
    {
        Destroy(gameObject);
    }
};
